chain/chain/src/store.rs

```diff
@@ -1815,116 +1835,211 @@ impl<'a> ChainStoreUpdate<'a> {
+            .expect("block data is not expected to be already cleaned")
+            .clone();
+        let height = block.header.inner_lite.height;
+        if height == self.get_genesis_height() {
+            // Broken GC prerequisites found
+            assert!(!is_fork);
+            return Ok(());
+        }
+
+        // 2. Delete shard_id-indexed data (shards, receipts, transactions)
+        for shard_id in 0..block.header.inner_rest.chunk_mask.len() as ShardId {
+            // 2a. Delete outgoing receipts (ColOutgoingReceipts)
+            store_update.delete(ColOutgoingReceipts, &get_block_shard_id(&block_hash, shard_id));
+            self.chain_store
+                .outgoing_receipts
+                .cache_remove(&get_block_shard_id(&block_hash, shard_id));
+            // 2b. Delete incoming receipts (ColIncomingReceipts)
+            store_update.delete(ColIncomingReceipts, &get_block_shard_id(&block_hash, shard_id));
+            self.chain_store
+                .incoming_receipts
+                .cache_remove(&get_block_shard_id(&block_hash, shard_id));
+            // 2c. Delete from chunk_hash_per_height_shard (ColChunkPerHeightShard)
+            store_update.delete(ColChunkPerHeightShard, &get_height_shard_id(height, shard_id));
+            self.chain_store
+                .chunk_hash_per_height_shard
+                .cache_remove(&get_height_shard_id(height, shard_id));
+            // 2d. Delete from next_block_with_new_chunk (ColNextBlockWithNewChunk)
+            store_update
+                .delete(ColNextBlockWithNewChunk, &get_block_shard_id(&block_hash, shard_id));
+            self.chain_store
+                .next_block_with_new_chunk
+                .cache_remove(&get_block_shard_id(&block_hash, shard_id));
+            // 2e. Delete from ColStateHeaders
+            let key = StateHeaderKey(shard_id, block_hash).try_to_vec()?;
+            store_update.delete(ColStateHeaders, &key);
+            // 2f. Delete from ColStateParts
+            // Already done, check chain.clear_downloaded_parts()
+        }
+        for chunk_header in block.chunks {
+            let (receipts, transactions) = match self.get_chunk_clone_from_header(&chunk_header) {
+                Ok(chunk) => (chunk.receipts, chunk.transactions),
+                _ => (vec![], vec![]),
+            };
+            // 2g. Delete from receipt_id_to_shard_id (ColReceiptIdToShardId)
+            for receipt in receipts {
+                store_update.delete(ColReceiptIdToShardId, receipt.receipt_id.as_ref());
+                self.chain_store.receipt_id_to_shard_id.cache_remove(&receipt.receipt_id.into());
+            }
+            // 2h. Delete from ColTransactions
+            for transaction in transactions {
+                store_update.delete(ColTransactions, transaction.get_hash().as_ref());
+                self.chain_store.transactions.cache_remove(&transaction.get_hash().into());
+            }
```

Comment:
Instead of creating empty vecs, I would just use `if let Ok(chunk) ...`.

Reply:
++